***Here we analyze the pre-existing data***

**Import Libraries and files**

In [1]:
import pandas as pd
import numpy as np

In [20]:
# Load the CSV file
lists = pd.read_csv("csvs/lists_2019_8.csv",encoding="ISO-8859-1")
commits = pd.read_csv("csvs/modified_commits_2019_8.csv",encoding="ISO-8859-1")
people = pd.read_csv("csvs/people_2019_8.csv",encoding="ISO-8859-1")
pr = pd.read_csv("csvs/pull-requests.csv",encoding="ISO-8859-1")
alias=pd.read_csv("csvs/modified_aliases_2019_8.csv",encoding="ISO-8859-1")

/tmp/ipykernel_257416/1690718824.py:3: DtypeWarning: Columns (14) have mixed types. Specify dtype option on import or set low_memory=False.
  commits = pd.read_csv("csvs/modified_commits_2019_8.csv",encoding="ISO-8859-1")


In [4]:
print(lists.shape)
print(commits.shape)
print(people.shape)
print(pr.shape)
print(alias.shape)

(312, 15)
(3655492, 17)
(33895, 4)
(569448, 12)
(47078, 5)


In [ ]:
# Search a string in a dataframe column
# lists['listname'].str.contains('abdera', case=False, na=False).any()

In [ ]:
# Counting unique values
# pr['state'].value_counts()

state
closed    558083
open       11365
Name: count, dtype: int64

In [ ]:
# Check any duplicate values
# pr['number'].duplicated().any()

In [ ]:
# Print the duplicate row values
# pr[pr['number'].duplicated()]

In [ ]:
# filtered_df = pr[pr['user'].str.contains('Darinj', case=False, na=False)]
# print(filtered_df)

In [ ]:
# # Find rows a certain column has certain values
# filtered_df = commits[commits['committerid'].isin([13521,25026,31916])]
# print(filtered_df)

In [ ]:
# # Filter rows where the a column has NaN values
# rows_with_nan = people[people['name'].isna()]
# print(rows_with_nan)

In [13]:
# Find rows where a certain column has certain values
filtered_df = commits[commits['authorid'].isin([20764])]
len(filtered_df)

334

In [14]:
filtered_df

list                                        messageid  committerid  \
1592568   128  1c989a81a0404e8bb14d7499d24115b1@git.apache.org        20764   
1592621   128  fada720ec9504b2188ed2c0c3482dcec@git.apache.org        20764   
1592630   128  e4f3ce8cff8f4e2ebbabe2c7c8dc2afa@git.apache.org        20764   
1593110   128  dd6ca2aa0df341ffac9464dab9f89502@git.apache.org        20764   
1593589   128  76107c99e52a43af939ceb183084d6cb@git.apache.org        20764   
...       ...                                              ...          ...   
1602433   128  a566f37c3af64f1180ddebf40227848d@git.apache.org        20764   
1602487   128  8dbbb64137f7493e917a3a899ecfb805@git.apache.org        20764   
1602488   128  8dbbb64137f7493e917a3a899ecfb805@git.apache.org        20764   
1602521   128  ea1d2428ad164302a74e560cbbe07c30@git.apache.org        20764   
1602522   128  ea1d2428ad164302a74e560cbbe07c30@git.apache.org        20764   

                commit_datetime file_operation  \
1592568  2015-05-19 03:06:32-07            mod   
1592621  2015-05-19 04:02:26-07            mod   
1592630  2015-05-19 04:28:56-07            mod   
1593110  2015-05-19 11:36:40-07            mod   
1593589  2015-05-20 05:45:33-07            add   
...                         ...            ...   
1602433  2015-05-27 05:19:41-07            mod   
1602487  2015-05-27 11:18:54-07            mod   
1602488  2015-05-27 11:18:54-07            mod   
1602521  2015-05-27 11:19:42-07            mod   
1602522  2015-05-27 11:19:42-07            mod   

                                                 file_name  addlines  \
1592568  git-wip-us.apache.org/repos/asf/incubator-igni...         7   
1592621  git-wip-us.apache.org/repos/asf/incubator-igni...        17   
1592630  git-wip-us.apache.org/repos/asf/incubator-igni...         4   
1593110  git-wip-us.apache.org/repos/asf/incubator-igni...        50   
1593589  git-wip-us.apache.org/repos/asf/incubator-igni...       164   
...                                                    ...       ...   
1602433  git-wip-us.apache.org/repos/asf/incubator-igni...        28   
1602487  git-wip-us.apache.org/repos/asf/incubator-igni...         1   
1602488  git-wip-us.apache.org/repos/asf/incubator-igni...        46   
1602521  git-wip-us.apache.org/repos/asf/incubator-igni...        16   
1602522  git-wip-us.apache.org/repos/asf/incubator-igni...       116   

         dellines      id  file_id  authorid         author_datetime  \
1592568         3  297629    86210     20764  2015-05-19 03:06:32-07   
1592621         0  297681    86262     20764  2015-05-19 04:02:26-07   
1592630         0  297689    86268     20764  2015-05-19 04:28:56-07   
1593110        50  298172    86466     20764  2015-05-19 11:36:40-07   
1593589         0  298745    86781     20764  2015-05-20 05:45:33-07   
...           ...     ...      ...       ...                     ...   
1602433         8  307492    88111     20764  2015-05-27 05:19:41-07   
1602487         1  307548    88219     20764  2015-05-27 11:18:54-07   
1602488        27  307549    88220     20764  2015-05-27 11:18:54-07   
1602521         2  307582    88217     20764  2015-05-27 11:19:42-07   
1602522        47  307583    88218     20764  2015-05-27 11:19:42-07   

         c_peopleid  a_peopleid                                sha_or_rev  \
1592568         NaN         NaN  b58c6f9e1be1ad3d1c7e0d7364a1086783eb0fb4   
1592621         NaN         NaN  a916dac74be025de73e6f0c9fe2b429fbef603c5   
1592630         NaN         NaN  41915f5465e13511171e34128ee81928ab89de5a   
1593110         NaN         NaN  0bd55632013b81d1bbf10f8da91d57e634c058f5   
1593589         NaN         NaN  3cc3fbe3168acaf49e691ffe74d780239968b311   
...             ...         ...                                       ...   
1602433         NaN         NaN  4be517c868e6798876a6eb44fcbc307ba4451603   
1602487         NaN         NaN  80f4ac2a3e31c681732cb144e14a614f6b8cb963   
1602488         NaN         NaN  80f

*Checking projects*

In [5]:
lists.columns

Index(['listid', 'listname', 'pj_alias', 'status', 'start_date', 'end_date',
       'dev_is_available', 'commit_is_available', 'is_in_attic', 'sponsor',
       'intro', 'istoplv', 'start_from_scratch', 'pj_url', 'pj_github_url'],
      dtype='object')

In [9]:
# Determining the total Unique values in a column
unique_values = lists['status'].unique()
print(unique_values)
print(unique_values.size)

[0 1 2]
3


In [10]:
# Count occurrences of each unique value in a column
unique_value_counts = lists['status'].value_counts()

print(unique_value_counts)

status
1    204
2     60
0     48
Name: count, dtype: int64


In [24]:
# Count occurrences of each unique value in a column
unique_value_counts = lists['commit_is_available'].value_counts()

print(unique_value_counts)

commit_is_available
t    276
f     36
Name: count, dtype: int64


In [ ]:
# Check if a column contains NaN
has_nan = lists['is_in_attic'].isna().any()

print(has_nan) 

False


*Checking commits*

In [26]:
commits.columns

Index(['list', 'messageid', 'committerid', 'commit_datetime', 'file_operation',
       'file_name', 'addlines', 'dellines', 'id', 'file_id', 'authorid',
       'author_datetime', 'c_peopleid', 'a_peopleid', 'sha_or_rev', 'format',
       'datetime'],
      dtype='object')

**The following code segment proves that some commit shas are missing**

In [ ]:
# Check if a column contains NaN
has_nan = commits['commit_datetime'].isna().any()

print(has_nan)

False


**The following code segment proves that we have commits for only 263 projects**

In [28]:
# Determining the total Unique values in a column
unique_values = commits['list'].unique()
print(unique_values)
print(unique_values.size)

[ 54 256 132  55 129 128 276 254   1  56   2  58  59  49 277  65 130  61
  62  63  64 257   3  70  67  68  69 259 260  81  71  66  72  73  74  75
  76  77  78  79  80 179  83 261  82 263  92 264  84   5  85  86  87  88
   6  89  90  91  13  93  94  95  96   8  10  97 267  11  98  99 100  12
 106 113  14 107 108 109 110 112 268 101 102 103 104 105  15 121  57 116
 269 270 176 117 177 178 181 118 183 184 185 186 188 291  52 190 119  29
 292 191 192 195  50 196 197 135 136 120 137 139  51 279 142 280 143 145
  21 283 154 156 157 159 160 161 163 285 164  24  25  26 168 169 286 287
 170 171 172 173 289 115 122 131 223 199 295 200 296 201 123 203 271  33
 297 124 204 205 207 208 209 210  37 298 211 272  16 273  17 212 213 214
 215 216 217 218 127 219 301 253  53  39 225 302  41 229 230 231 232  43
 235 303 238 240 241 242 243 244 306 246 248 250 308 310 252 312  18  20
 153 288  30  31  45  48 134 138 141 281 144 146 155  23 167 175  27 182
 290 189 187  28 194 293 294 206 202  32  34  35 29

In [61]:
# Count occurrences of each unique value in a column
unique_value_counts = commits['format'].value_counts()

print(unique_value_counts)

format
git    2221638
svn    1433854
Name: count, dtype: int64


In [31]:
# Determining the total Unique values in a column
unique_values = commits['sha_or_rev'].unique()
print(unique_values)
print(unique_values.size)

[nan 'aa01cd4e9a60507bafb98707c7793c3a89424db8' '30857' ... 34380.0
 34392.0 34393.0]
293404


*Checking people*

In [32]:
people.columns

Index(['personid', 'name', 'isdev', 'hasemail'], dtype='object')

In [35]:
people.head(10)

personid                  name isdev hasemail
0         1      Benjamin Hindman     f        f
1         2  Lewis John McGibbney     f        f
2         3         Andy Seaborne     f        f
3         4          Brett Porter     f        f
4         5          Olivier Lamy     f        f
5         6    Jarek Jarcec Cecho     f        f
6         7         Matei Zaharia     f        f
7         8            Joe Bowser     f        f
8         9          Ross Gardler     f        f
9        10          Brock Noland     f        f

**Not all persons are developers and they may not have commits**

In [34]:
# Determining the total Unique values in a column
unique_values = people['personid'].unique()
print(unique_values)
print(unique_values.size)

[    1     2     3 ... 33897 33898 33899]
33895


In [37]:
# Determining the total Unique values in a column
unique_values = people['isdev'].unique()
print(unique_values)
print(unique_values.size)

['f']
1


*Checking alias*

In [39]:
alias.columns

Index(['aliasid', 'email', 'name', 'personid', 'from_commit'], dtype='object')

In [ ]:
alias.shape

In [41]:
alias.head(10)

aliasid                                              email  \
0    22233                                    jira@apache.org   
1    12937  reply+i-2355594-3a189c6b1c866515f1dd1db09f4f45...   
2    12967                                    jira@apache.org   
3    24428                            matei@eecs.berkeley.edu   
4    24451                                    jira@apache.org   
5     1120                         rgardler@opendirective.com   
6    16699                                    jira@apache.org   
7     1144                                    jira@apache.org   
8    16854                                    jira@apache.org   
9    34265                     matthieu@yahoo-inc.com.INVALID   

                                    name  personid from_commit  
0        Andy Seaborne (Assigned) (JIRA)         3           f  
1                           Olivier Lamy         5           f  
2         Olivier Lamy (Reopened) (JIRA)         5           f  
3                          Matei Zaharia         7           f  
4       Matei Zaharia (Commented) (JIRA)         7           f  
5                           Ross Gardler         9           f  
6          Brock Noland (Created) (JIRA)        10           f  
7  Lahiru Gunathilake (Commented) (JIRA)        12           f  
8           Mike Percy (Assigned) (JIRA)        14           f  
9                         Matthieu Morel        15           f

In [44]:
# Determining the total Unique values in a column
unique_values = alias['aliasid'].unique()
print(unique_values)
print(unique_values.size)

[22233 12937 12967 ...  8569 20573 33862]
47078


In [ ]:
# Convert both arrays to lowercase for case-insensitive matching
lowercase_ndarray = np.array(unique_values)
lowercase_column = commits["authorid"]

# Convert to set for faster lookup
lowercase_column_set = set(lowercase_column)

# Find strings present in the DataFrame
present_strings = [s for s in lowercase_ndarray if s in lowercase_column_set]

# Find strings that were NOT found
not_found_strings = [s for s in lowercase_column_set if s not in  lowercase_ndarray]

# Filter the DataFrame for only the matched strings
matched_df = commits[commits["authorid"].isin(set(present_strings))]

# # Count occurrences of 0, 1, and 2 in the "status" column
# status_counts = matched_df["status"].value_counts().to_dict()

# Print results
print("Strings present in DataFrame:", present_strings)
print("Total strings present:", len(present_strings))
# print("Status counts:", status_counts)

# # Print count of each status (default to 0 if not found)
# print(f"Count of 0: {status_counts.get(0, 0)}")
# print(f"Count of 1: {status_counts.get(1, 0)}")
# print(f"Count of 2: {status_counts.get(2, 0)}")

# Print the names not found and their count
print("Strings NOT found in DataFrame:", not_found_strings)
print("Total strings NOT found:", len(not_found_strings))

In [43]:
# Determining the total Unique values in a column
unique_values = alias['personid'].unique()
print(unique_values)
print(unique_values.size)

[    3     5     7 ... 15113 15122 15155]
33629


In [45]:
# Determining the total Unique values in a column
unique_values = alias['from_commit'].unique()
print(unique_values)
print(unique_values.size)

['f' 't']
2


In [46]:
# Count occurrences of each unique value in a column
unique_value_counts = alias['from_commit'].value_counts()

print(unique_value_counts)

from_commit
f    36288
t    10790
Name: count, dtype: int64


In [ ]:
alias.shape

*Checking Pull Requests*

In [8]:
pr.columns

Index(['repo', 'number', 'user', 'user_type', 'state', 'created_at',
       'updated_at', 'closed_at', 'merged_at', 'merge_commit_sha', 'issue_url',
       'body'],
      dtype='object')

**This code snippet says that pull requests are available from 203 projects**

In [25]:
# Determining the total Unique values in a column
unique_values = pr['repo'].unique()
print(unique_values)
print(unique_values.size)

['Abdera' 'Accumulo' 'ACE' 'ActiveMQ' 'activemq-apollo' 'Airavata'
 'Airflow' 'Allura' 'Ambari' 'ant-easyant-core' 'ant-ivy' 'Any23'
 'apex-core' 'Aries' 'AsterixDB' 'Atlas' 'Aurora' 'Beam' 'Bigtop'
 'Bloodhound' 'Brooklyn' 'Buildr' 'bval' 'Calcite' 'CarbonData'
 'Cassandra' 'Cayenne' 'Celix' 'Chukwa' 'Clerezza' 'climate' 'CloudStack'
 'commons-ognl' 'Cordova' 'CouchDB' 'creadur-rat' 'Crunch' 'cTAKES'
 'Curator' 'CXF' 'DataFu' 'db-jdo' 'Deltacloud' 'DeltaSpike' 'Derby'
 'DirectMemory' 'directory-server' 'DistributedLog' 'Drill' 'Dubbo'
 'Eagle' 'Empire-db' 'ESME' 'Falcon' 'Felix' 'Fineract' 'flex-sdk' 'Flink'
 'Flume' 'Fluo' 'FreeMarker' 'FtpServer' 'Geode' 'Geronimo'
 'geronimo-yoko' 'Giraph' 'Gora' 'Griffin' 'Groovy' 'guacamole-client'
 'Hama' 'HAWQ' 'HCatalog' 'Helix' 'Ignite' 'Impala' 'Amaterasu'
 'Annotator' 'ariatosca' 'BatchEE' 'BRPC' 'Crail' 'Daffodil'
 'DataSketches' 'DLab' 'DolphinScheduler' 'Doris' 'Druid' 'ECharts'
 'Edgent' 'Flagon' 'Gobblin' 'Heron' 'Hivemall' 'Hudi' 'Ice

In [ ]:
# # Check if 'project_name' contains 'incubator-' and count occurrences
# p = pr['repo'].str.contains(r'^retired-', na=False).unique()

# # Count the number of rows that contain 'incubator-'
# count = p.sum()
# print(count)

In [23]:
p=pr['repo'].unique()
p=pd.DataFrame(p)
p.to_csv("pr.csv",index=False)

**Among 252 projects in PR Lists, 62 have "incubator-" written in front of them which is causing string mismatch with project names in Lists
Again, 15 among the 62 have "retired-" written after "incubator-". We need to remove them both.**

In [22]:
# Remove "incubator-" from the start of strings
pr["repo"] = pr["repo"].str.replace(r"^incubator-", "", regex=True)
pr["repo"] = pr["repo"].str.replace(r"^retired-", "", regex=True)

In [26]:
lowercase_ndarray = np.array([str(s).lower() for s in unique_values])
lowercase_column = lists['listname'].str.lower()

# Find strings that are present in the DataFrame column
present_strings = [s for s in lowercase_ndarray if s in lowercase_column.values]

# Print the matching strings
print("Strings present in DataFrame:", present_strings)

# Print the total count
print("Total strings present:", len(present_strings))


Strings present in DataFrame: ['abdera', 'accumulo', 'ace', 'activemq', 'airavata', 'airflow', 'allura', 'ambari', 'any23', 'aries', 'asterixdb', 'atlas', 'aurora', 'beam', 'bigtop', 'bloodhound', 'brooklyn', 'buildr', 'calcite', 'carbondata', 'cassandra', 'cayenne', 'celix', 'chukwa', 'clerezza', 'cloudstack', 'cordova', 'couchdb', 'crunch', 'ctakes', 'curator', 'cxf', 'datafu', 'deltacloud', 'deltaspike', 'derby', 'directmemory', 'distributedlog', 'drill', 'dubbo', 'eagle', 'empire-db', 'esme', 'falcon', 'felix', 'fineract', 'flink', 'flume', 'fluo', 'freemarker', 'ftpserver', 'geode', 'geronimo', 'giraph', 'gora', 'griffin', 'groovy', 'hama', 'hawq', 'hcatalog', 'helix', 'ignite', 'impala', 'amaterasu', 'annotator', 'ariatosca', 'batchee', 'brpc', 'crail', 'daffodil', 'datasketches', 'dlab', 'dolphinscheduler', 'doris', 'druid', 'echarts', 'edgent', 'flagon', 'gobblin', 'heron', 'hivemall', 'hudi', 'iceberg', 'iotdb', 'livy', 'milagro', 'mxnet', 'myriad', 'nemo', 'omid', 'pinot', 'r

**We finally got 21 projects are unsustainable. 155 are sustainable. 42 in_incubation**

In [27]:
# Filter the DataFrame for only the matched strings
matched_df = lists[lists["listname"].str.lower().isin(present_strings)]

# Count occurrences of 0, 1, and 2 in the "status" column
status_counts = matched_df["status"].value_counts().to_dict()

# Print results
print("Strings present in DataFrame:", present_strings)
print("Total strings present:", len(present_strings))
print("Status counts:", status_counts)

# Print count of each status (default to 0 if not found)
print(f"Count of 0: {status_counts.get(0, 0)}")
print(f"Count of 1: {status_counts.get(1, 0)}")
print(f"Count of 2: {status_counts.get(2, 0)}")

Strings present in DataFrame: ['abdera', 'accumulo', 'ace', 'activemq', 'airavata', 'airflow', 'allura', 'ambari', 'any23', 'aries', 'asterixdb', 'atlas', 'aurora', 'beam', 'bigtop', 'bloodhound', 'brooklyn', 'buildr', 'calcite', 'carbondata', 'cassandra', 'cayenne', 'celix', 'chukwa', 'clerezza', 'cloudstack', 'cordova', 'couchdb', 'crunch', 'ctakes', 'curator', 'cxf', 'datafu', 'deltacloud', 'deltaspike', 'derby', 'directmemory', 'distributedlog', 'drill', 'dubbo', 'eagle', 'empire-db', 'esme', 'falcon', 'felix', 'fineract', 'flink', 'flume', 'fluo', 'freemarker', 'ftpserver', 'geode', 'geronimo', 'giraph', 'gora', 'griffin', 'groovy', 'hama', 'hawq', 'hcatalog', 'helix', 'ignite', 'impala', 'amaterasu', 'annotator', 'ariatosca', 'batchee', 'brpc', 'crail', 'daffodil', 'datasketches', 'dlab', 'dolphinscheduler', 'doris', 'druid', 'echarts', 'edgent', 'flagon', 'gobblin', 'heron', 'hivemall', 'hudi', 'iceberg', 'iotdb', 'livy', 'milagro', 'mxnet', 'myriad', 'nemo', 'omid', 'pinot', 'r

In [ ]:
# # Convert both arrays to lowercase for case-insensitive matching
# lowercase_ndarray = np.array([str(s).lower() for s in unique_values])
# lowercase_column = lists["listname"].str.lower()

# # Convert to set for faster lookup
# lowercase_column_set = set(lowercase_column)

# # Find strings present in the DataFrame
# present_strings = [s for s in lowercase_ndarray if s in lowercase_column_set]

# # Find strings that were NOT found
# not_found_strings = [s for s in lowercase_column_set if s not in  lowercase_ndarray]

# # Filter the DataFrame for only the matched strings
# matched_df = lists[lists["listname"].str.lower().isin(set(present_strings))]

# # Count occurrences of 0, 1, and 2 in the "status" column
# status_counts = matched_df["status"].value_counts().to_dict()

# # Print results
# print("Strings present in DataFrame:", present_strings)
# print("Total strings present:", len(present_strings))
# print("Status counts:", status_counts)

# # Print count of each status (default to 0 if not found)
# print(f"Count of 0: {status_counts.get(0, 0)}")
# print(f"Count of 1: {status_counts.get(1, 0)}")
# print(f"Count of 2: {status_counts.get(2, 0)}")

# # Print the names not found and their count
# print("Strings NOT found in DataFrame:", not_found_strings)
# print("Total strings NOT found:", len(not_found_strings))

Strings present in DataFrame: ['abdera', 'accumulo', 'ace', 'activemq', 'airavata', 'airflow', 'allura', 'ambari', 'any23', 'aries', 'asterixdb', 'atlas', 'aurora', 'beam', 'bigtop', 'bloodhound', 'brooklyn', 'buildr', 'calcite', 'carbondata', 'cassandra', 'cayenne', 'celix', 'chukwa', 'clerezza', 'cloudstack', 'cordova', 'couchdb', 'crunch', 'ctakes', 'curator', 'cxf', 'datafu', 'deltacloud', 'deltaspike', 'derby', 'directmemory', 'distributedlog', 'drill', 'dubbo', 'eagle', 'empire-db', 'esme', 'falcon', 'felix', 'fineract', 'flink', 'flume', 'fluo', 'freemarker', 'ftpserver', 'geode', 'geronimo', 'giraph', 'gora', 'griffin', 'groovy', 'hama', 'hawq', 'hcatalog', 'helix', 'ignite', 'impala', 'amaterasu', 'annotator', 'batchee', 'brpc', 'crail', 'daffodil', 'datasketches', 'dlab', 'dolphinscheduler', 'doris', 'druid', 'echarts', 'edgent', 'flagon', 'gobblin', 'heron', 'hivemall', 'hudi', 'iceberg', 'iotdb', 'livy', 'milagro', 'mxnet', 'myriad', 'nemo', 'omid', 'pinot', 'ratis', 'rya',

In [ ]:
# name = pr["repo"].dropna().unique()  # Get unique non-null values
# name = pd.DataFrame(name, columns=["repo"])
# name.to_csv("pr_repo.csv", index=False)

In [ ]:
# name=lists[['listname','pj_alias']]
# name = name.dropna()
# name.to_csv("prj_name.csv",index=False)